### ISSUE TITLE
train.py - saving model as checkpoint vs. model?

### ISSUE THREAD

**Issue by bricksaver** [pranjaldatta/DenseDepth-Pytorch#6]
Hi! I had a quick question. The train.py code saves the model as a checkpoint and test.py loads the model as a checkpoint. Is there any advantage/disadvantage to saving/using a model as a "checkpoint" rather than a "model?

Are checkpoints saveable?

**Comment by pranjaldatta** [comment]
> Hi! I had a quick question. The train.py code saves the model as a checkpoint and test.py loads the model as a checkpoint. Is there any advantage/disadvantage to saving/using a model as a "checkpoint" rather than a "model?
> 
> Are checkpoints saveable?

Okay so I guess, just to rephrase the question a bit, I think you are basically asking *why only store the state_dict() of the model and not the entire model?* (please let me know if I misunderstood the question).

The reasoning is as follows, 
* If you save the entire model, (i.e. `torch.save(model)` ), when you load up the model later, PT expects a similar project structure. Meaning,  PT would throw an error if you try loading up the model in a different environment (a different project with different files, folder structure, etc), which kinda hampers greater usability (again, my personal opinion). Because it's kinda limiting if users have to limit their applications to this project alone.

* For example, lets say you trained the model and you want to use the model in another project of yours, then saving up an entire model would mean the folder `densedepth` would have to copied as it in your project (lets not even get into the potential import hell).

* On the contrary, saving just the state_dict() gives you the freedom to instantiate the model anywhere you want and load the weights in whichever environment that suits your application needs. As you can see [here](https://github.com/pranjaldatta/DenseDepth-Pytorch/blob/820fabe83e32f7fceebfadfc373af43118241f54/densedepth/test.py#L37) test.py is instantiating the model + loading the weights. 

* Just a note, I personally think this more convenient. Please feel free to change the code if it suits your needs. 

Another question might be, *why save optimizer state dicts etc?*, Well, that's just for training purposes nothing more!


I hope this answers your question. Please feel free to clarify/suggest any alternatives.
